File: Engine/Sources/CardCopilotEngine/Models/CategoryTaxonomy.swift
import Foundation

/// The canonical vocabulary shared by purchase contexts, catalogue predicates, and UI entry
/// points. Display labels are allowed to be friendly; values sent to the engine are not.
///
/// This is intentionally a small boundary normalizer rather than a permissive fuzzy matcher.
/// Known aliases are collapsed to the catalogue's exact ids, while unknown values are preserved
/// so a new catalogue category is not silently misclassified.
public enum CategoryTaxonomy {

    /// Real purchase/merchant categories. Predicate-only dimensions such as `recurring` and
    /// `ownerSelectedCategory` are deliberately excluded: no database row should claim that a
    /// merchant *is* one of those conditions.
    public static let purchaseCategoryIDs = Set(registry.categories.map(\.id))

    /// Tokens understood by catalogue matching that are not valid merchant classifications.
    public static let ruleSideCategoryIDs = Set(registry.ruleSideCategories.map(\.id))

    private static let registry = SeedLoader.purchaseCategories

    private static let definitionsByID: [String: PurchaseCategoryDefinition] = {
        let definitions = registry.categories + registry.ruleSideCategories
        let pairs = definitions.map { ($0.id, $0) }
        let result = Dictionary(pairs, uniquingKeysWith: { first, duplicate in
            preconditionFailure("Duplicate category id in purchase-categories.json: \(duplicate.id)")
        })
        precondition(result.count == definitions.count,
                     "purchase-categories.json contains duplicate ids")
        return result
    }()

    private static let canonicalIDByCompactKey: [String: String] = {
        var result: [String: String] = [:]
        for definition in registry.categories + registry.ruleSideCategories {
            for raw in [definition.id] + definition.aliases {
                let key = compactKey(raw)
                if let existing = result[key] {
                    precondition(existing == definition.id,
                                 "Category alias '\(raw)' is ambiguous: \(existing), \(definition.id)")
                }
                result[key] = definition.id
            }
        }
        return result
    }()

    /// Converts a display/input alias to the catalogue's canonical category id.
    public static func canonicalID(_ raw: String) -> String {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return trimmed }

        return canonicalIDByCompactKey[compactKey(trimmed)] ?? trimmed
    }

    /// Canonicalizes and validates a value intended for a persisted purchase/merchant field.
    public static func canonicalPurchaseID(_ raw: String) -> String? {
        let canonical = canonicalID(raw)
        return purchaseCategoryIDs.contains(canonical) ? canonical : nil
    }

    /// Shared human-readable fallback for every non-visual surface.
    public static func displayName(for raw: String) -> String {
        let category = canonicalID(raw)
        if let displayName = definitionsByID[category]?.displayName { return displayName }
        guard !category.isEmpty else { return category }
        var spaced = ""
        for character in category {
            if character.isUppercase, !spaced.isEmpty { spaced.append(" ") }
            spaced.append(character)
        }
        return spaced.prefix(1).uppercased() + spaced.dropFirst().lowercased()
    }

    private static func compactKey(_ raw: String) -> String {
        raw.folding(options: [.diacriticInsensitive, .caseInsensitive],
                    locale: Locale(identifier: "en_CA"))
            .unicodeScalars
            .filter { CharacterSet.alphanumerics.contains($0) }
            .map(String.init)
            .joined()
            .lowercased()
    }
}

File: android/core/engine/src/main/kotlin/com/cardcopilot/engine/models/CategoryTaxonomy.kt
package com.cardcopilot.engine.models

import com.cardcopilot.engine.loading.SeedLoader
import java.text.Normalizer
import java.util.Locale
import kotlinx.serialization.Serializable

@Serializable
data class PurchaseCategoryDefinition(
    val id: String,
    val displayName: String,
    val parentID: String? = null,
    val merchantGroupID: String? = null,
    val aliases: List<String> = emptyList(),
)

@Serializable
data class PurchaseCategoryRegistry(
    val taxonomyVersion: String = "1.0",
    val categories: List<PurchaseCategoryDefinition> = emptyList(),
    val ruleSideCategories: List<PurchaseCategoryDefinition> = emptyList(),
)

/** Kotlin twin of Swift's canonical purchase-category boundary. */
object CategoryTaxonomy {
    private val registry by lazy(SeedLoader::loadPurchaseCategories)

    val purchaseCategoryIds: Set<String> by lazy { registry.categories.mapTo(mutableSetOf()) { it.id } }

    val taxonomyVersion: String get() = registry.taxonomyVersion

    val ruleSideCategoryIds: Set<String> by lazy {
        registry.ruleSideCategories.mapTo(mutableSetOf()) { it.id }
    }

    private val definitionsById: Map<String, PurchaseCategoryDefinition> by lazy {
        val definitions = registry.categories + registry.ruleSideCategories
        definitions.associateBy { it.id }.also {
            require(it.size == definitions.size) {
                "purchase-categories.json contains duplicate ids"
            }
        }
    }

    private val canonicalIdByCompactKey: Map<String, String> by lazy {
        buildMap {
            for (definition in registry.categories + registry.ruleSideCategories) {
                for (raw in listOf(definition.id) + definition.aliases) {
                    val key = compactKey(raw)
                    val existing = this[key]
                    require(existing == null || existing == definition.id) {
                        "Category alias '$raw' is ambiguous: $existing, ${definition.id}"
                    }
                    put(key, definition.id)
                }
            }
        }
    }

    fun canonicalId(raw: String): String {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return trimmed
        return canonicalIdByCompactKey[compactKey(trimmed)] ?: trimmed
    }

    fun canonicalPurchaseId(raw: String): String? =
        canonicalId(raw).takeIf(purchaseCategoryIds::contains)

    fun parentIds(raw: String): List<String> {
        val result = mutableListOf<String>()
        val visited = mutableSetOf<String>()
        var current = canonicalId(raw)
        while (true) {
            val parent = definitionsById[current]?.parentID ?: break
            if (!visited.add(parent)) break
            result += parent
            current = parent
        }
        return result
    }

    fun merchantGroupId(raw: String): String? = definitionsById[canonicalId(raw)]?.merchantGroupID

    fun displayName(raw: String): String {
        val category = canonicalId(raw)
        return definitionsById[category]?.displayName ?: humanize(category)
    }

    private fun compactKey(raw: String): String = Normalizer
        .normalize(raw, Normalizer.Form.NFD)
        .replace("\\p{InCombiningDiacriticalMarks}+".toRegex(), "")
        .filter(Char::isLetterOrDigit)
        .lowercase(Locale.CANADA)

    private fun humanize(category: String): String {
        if (category.isEmpty()) return category
        val spaced = buildString {
            category.forEach { character ->
                if (character.isUpperCase() && isNotEmpty()) append(' ')
                append(character)
            }
        }
        return spaced.replaceFirstChar { it.uppercase(Locale.CANADA) }.lowercase()
            .replaceFirstChar { it.uppercase(Locale.CANADA) }
    }
}
